src/python-bindings/htcondor2/schedd.cpp

```diff
@@ -1 +1,950 @@
+    switch( action ) {
+        case JA_HOLD_JOBS:
+            result = schedd.holdJobs(constraint, reason_string, reason_code, NULL, AR_TOTALS );
+            break;
+
+        case JA_RELEASE_JOBS:
+            result = schedd.releaseJobs(constraint, reason_string, NULL, AR_TOTALS );
+            break;
+
+        case JA_REMOVE_JOBS:
+            result = schedd.removeJobs(constraint, reason_string, NULL, AR_TOTALS );
+            break;
+
+        case JA_REMOVE_X_JOBS:
+            result = schedd.removeXJobs(constraint, reason_string, NULL, AR_TOTALS );
+            break;
+
+        case JA_VACATE_JOBS:
+        case JA_VACATE_FAST_JOBS:
+            {
+            auto vacate_type = action == JA_VACATE_JOBS ? VACATE_GRACEFUL : VACATE_FAST;
+            result = schedd.vacateJobs(constraint, vacate_type, NULL, AR_TOTALS);
+            } break;
+
+        case JA_SUSPEND_JOBS:
+            result = schedd.suspendJobs(constraint, reason_string, NULL, AR_TOTALS );
+            break;
+
+        case JA_CONTINUE_JOBS:
+            result = schedd.continueJobs(constraint, reason_string, NULL, AR_TOTALS );
+            break;
+
+        default:
+            // This was HTCondorEnumError, in version 1.
+            PyErr_SetString(PyExc_RuntimeError, "Job action not implemented.");
+            return NULL;
+    }
+
+    if( result == NULL ) {
+        // This was HTCondorReplyError, in version 1.
+        PyErr_SetString(PyExc_RuntimeError, "Error when performing action on the schedd.");
+        return NULL;
+    }
+
+
+    return py_new_classad2_classad(result->Copy());
+}
+
+
+static PyObject *
+_schedd_edit_job_ids(PyObject *, PyObject * args) {
+    // schedd_edit_job_ids(addr, job_list, attr, value, flags)
+
+    const char * addr = NULL;
+    const char * job_list = NULL;
+    const char * attr = NULL;
+    const char * value = NULL;
+    long flags = 0;
+
+    if(! PyArg_ParseTuple( args, "zzzzl", & addr, & job_list, & attr, & value, & flags )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    flags = flags | SetAttribute_NoAck;
+
+
+    // FIXME: It was not at all clear that ConnectionSentry opened the (?!)
+    // global connection to the schedd, but having it disconnect when it
+    // goes out of scope is really kind of nice.
+    //
+    // And yes, I know, the ConnectQ()/SetAttribute()/DisconnectQ() interface
+    // is brokenly stupid.
+    DCSchedd schedd(addr);
+    if( ConnectQ(schedd) == NULL ) {
+        // This was HTCondorIOError, in version 1.
+        PyErr_SetString(PyExc_IOError, "Failed to connect to schedd.");
+        return NULL;
+    }
+
+    StringList ids(job_list);
+
+
+    long matchCount = 0;
+    const char * id = NULL;
+    for( ids.rewind(); (id = ids.next()) != NULL; ) {
+        JOB_ID_KEY jobIDKey;
+        if(! jobIDKey.set(id)) {
+            // FIXME: Check error return and use error stack.
+            DisconnectQ(NULL);
```

Comment:
bad form to pass NULL here,  should be the Qmgr_connection you got back from ConnectQ